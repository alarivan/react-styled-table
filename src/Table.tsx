import * as React from 'react'
import * as R from 'ramda'
import styled, { StyledComponent } from 'styled-components'
import TableHeaderCell from './TableHeaderCell'
import { ReactStyledTableDataItem, ReactStyledTableColumn, SortOrder } from '.'

interface Props
  extends React.DetailedHTMLProps<
    React.HTMLAttributes<HTMLDivElement>,
    HTMLDivElement
  > {
  data: ReactStyledTableDataItem[]
  columns: ReactStyledTableColumn[]
  tableCell?: StyledComponent<'td', any, {}, never>
}

export const ReactStyledTableCell = styled.td`
  min-width: 140px;
`

export const ReactStyledTableContainer = styled.div`
  max-width: 500px;
  overflow-x: auto;
`

export const Table = ({ data, columns, tableCell, className }: Props) => {
  const [sortedData, setSortedData] = React.useState<
    ReactStyledTableDataItem[]
  >(data)
  const [sortByValue, setSortByValue] = React.useState<string>()
  const [sortOrder, setSortOrder] = React.useState<SortOrder>()

  const handleSortChange = (value: string) => {
    let nextSortOrder: SortOrder
    switch (sortOrder) {
      case 'ascend':
        nextSortOrder = 'descend'
        break
      case undefined:
        nextSortOrder = 'ascend'
        break
    }

    if (!nextSortOrder) {
      setSortedData(data)
    } else if (value) {
      const sorter = R[nextSortOrder](R.prop(value))
      setSortedData(R.sortWith([sorter], sortedData))
    }

    setSortByValue(value)
    setSortOrder(nextSortOrder)
  }

  const renderTableCell = (
    item: ReactStyledTableDataItem,
    column: ReactStyledTableColumn,
    index: number
  ) => {
    return React.createElement(
      column.tableCellComponent || tableCell || ReactStyledTableCell,
      { key: index },
      column.render ? column.render(item[column.value]) : item[column.value]
    )
  }

  return (
    <ReactStyledTableContainer className={className}>
      <table>
        <thead>
          <tr>
            {columns.map((column) => (
              <TableHeaderCell
                key={column.value}
                value={column.value}
                label={column.label}
                sortable={column.sortable}
                sortByValue={sortByValue}
                sortOrder={sortOrder}
                onClick={handleSortChange}
              />
            ))}
          </tr>
        </thead>
        <tbody>
          {sortedData.map((item, idx) => (
            <tr key={idx}>
              {columns.map((column, idx) => renderTableCell(item, column, idx))}
            </tr>
          ))}
        </tbody>
      </table>
    </ReactStyledTableContainer>
  )
}
