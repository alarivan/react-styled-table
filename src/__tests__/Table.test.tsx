import 'jest-styled-components'
import styled from 'styled-components'
import React from 'react'
import { render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { ReactStyledTableCell, Table } from '../Table'
import { ReactStyledTableColumn, ReactStyledTableDataItem } from '..'

const BlueTableCell = styled(ReactStyledTableCell)`
  background: lightblue;
`

const GreenTableCell = styled(ReactStyledTableCell)`
  background: green;
`

const data: ReactStyledTableDataItem[] = [
  {
    name: 'test-name-1',
    movie_rating: 4,
    city: 'Vancouver',
    country: 'Canada'
  },
  {
    name: 'test-name-2',
    movie_rating: 5,
    city: 'Frankfurt',
    country: 'Germany'
  }
]

const columns: ReactStyledTableColumn[] = [
  {
    label: 'Name',
    value: 'name',
    sortable: true
  },
  {
    label: 'Rating',
    value: 'movie_rating',
    render: (value: string | number) => {
      return `~${value}`
    }
  },
  {
    label: 'City',
    value: 'city'
  },
  {
    label: 'Country',
    value: 'country',
    tableCellComponent: BlueTableCell
  }
]

test('should render correctly with all possible data and columns variants', () => {
  const { container } = render(<Table data={data} columns={columns} />)

  screen.getByText('~5')

  expect(screen.getByText('Germany')).toHaveStyleRule('background: lightblue;')

  expect(container.firstChild).toMatchSnapshot()
})

test('should render correctly with empty data', () => {
  const { container } = render(<Table data={[]} columns={columns} />)

  expect(container.firstChild).toMatchSnapshot()
})

test('should render correctly with empty columns', () => {
  const { container } = render(<Table data={data} columns={[]} />)

  expect(container.firstChild).toMatchSnapshot()
})

test('should render correctly with empty columns and data', () => {
  const { container } = render(<Table data={[]} columns={[]} />)

  expect(container.firstChild).toMatchSnapshot()
})

test('should render correctly with custom table cell', () => {
  const { container } = render(
    <Table data={data} columns={columns} tableCell={GreenTableCell} />
  )

  expect(screen.getByText('Frankfurt')).toHaveStyleRule(
    'background: lightblue;'
  )
  expect(screen.getByText('Germany')).toHaveStyleRule('background: green;')
  expect(screen.getByText('test-name-1')).toHaveStyleRule('background: green;')
  expect(screen.getByText('~5')).toHaveStyleRule('background: green;')

  expect(container.firstChild).toMatchSnapshot()
})

test('should switch order', async () => {
  render(<Table data={data} columns={columns} />)

  const names = screen.queryAllByText(/test-name-[1-2]/)

  expect(names[0].textContent).toStrictEqual('test-name-1')
  expect(names[1].textContent).toStrictEqual('test-name-2')

  const sortableHeaderCell = screen.getByText(/Name/)

  userEvent.click(sortableHeaderCell)

  expect(names[0].textContent).toStrictEqual('test-name-1')
  expect(names[1].textContent).toStrictEqual('test-name-2')

  userEvent.click(sortableHeaderCell)

  expect(names[0].textContent).toStrictEqual('test-name-2')
  expect(names[1].textContent).toStrictEqual('test-name-1')
})
