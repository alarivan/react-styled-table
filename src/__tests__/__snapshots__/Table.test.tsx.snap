// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly with all possible data and columns variants 1`] = `
.c1 {
  cursor: pointer;
}

.c2 {
  min-width: 140px;
}

.c0 {
  max-width: 500px;
  overflow-x: auto;
}

.c3 {
  background: lightblue;
}

<div
  class="c0"
>
  <table>
    <thead>
      <tr>
        <th
          class="c1"
          role="button"
        >
          Name
          <span>
            ↑
          </span>
           
          <span>
            ↓
          </span>
        </th>
        <th
          class=""
        >
          Rating
        </th>
        <th
          class=""
        >
          City
        </th>
        <th
          class=""
        >
          Country
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td
          class="c2"
        >
          test-name-1
        </td>
        <td
          class="c2"
        >
          ~4
        </td>
        <td
          class="c2"
        >
          Vancouver
        </td>
        <td
          class="c2 c3"
        >
          Canada
        </td>
      </tr>
      <tr>
        <td
          class="c2"
        >
          test-name-2
        </td>
        <td
          class="c2"
        >
          ~5
        </td>
        <td
          class="c2"
        >
          Frankfurt
        </td>
        <td
          class="c2 c3"
        >
          Germany
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`should render correctly with custom table cell 1`] = `
.c1 {
  cursor: pointer;
}

.c2 {
  min-width: 140px;
}

.c0 {
  max-width: 500px;
  overflow-x: auto;
}

.c4 {
  background: lightblue;
}

.c3 {
  background: green;
}

<div
  class="c0"
>
  <table>
    <thead>
      <tr>
        <th
          class="c1"
          role="button"
        >
          Name
          <span>
            ↑
          </span>
           
          <span>
            ↓
          </span>
        </th>
        <th
          class=""
        >
          Rating
        </th>
        <th
          class=""
        >
          City
        </th>
        <th
          class=""
        >
          Country
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td
          class="c2 c3"
        >
          test-name-1
        </td>
        <td
          class="c2 c3"
        >
          ~4
        </td>
        <td
          class="c2 c3"
        >
          Vancouver
        </td>
        <td
          class="c2 c4"
        >
          Canada
        </td>
      </tr>
      <tr>
        <td
          class="c2 c3"
        >
          test-name-2
        </td>
        <td
          class="c2 c3"
        >
          ~5
        </td>
        <td
          class="c2 c3"
        >
          Frankfurt
        </td>
        <td
          class="c2 c4"
        >
          Germany
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`should render correctly with empty columns 1`] = `
.c0 {
  max-width: 500px;
  overflow-x: auto;
}

<div
  class="c0"
>
  <table>
    <thead>
      <tr />
    </thead>
    <tbody>
      <tr />
      <tr />
    </tbody>
  </table>
</div>
`;

exports[`should render correctly with empty columns and data 1`] = `
.c0 {
  max-width: 500px;
  overflow-x: auto;
}

<div
  class="c0"
>
  <table>
    <thead>
      <tr />
    </thead>
    <tbody />
  </table>
</div>
`;

exports[`should render correctly with empty data 1`] = `
.c1 {
  cursor: pointer;
}

.c0 {
  max-width: 500px;
  overflow-x: auto;
}

<div
  class="c0"
>
  <table>
    <thead>
      <tr>
        <th
          class="c1"
          role="button"
        >
          Name
          <span>
            ↑
          </span>
           
          <span>
            ↓
          </span>
        </th>
        <th
          class=""
        >
          Rating
        </th>
        <th
          class=""
        >
          City
        </th>
        <th
          class=""
        >
          Country
        </th>
      </tr>
    </thead>
    <tbody />
  </table>
</div>
`;
